source/blog/2017-10-10-glimmer-progress-report.md

```diff
@@ -0,0 +1,645 @@
+this functionality is usually referred to as a portal, and as of React 16 is
+included by default in `react-dom`.)
+
+For example, if I had a `Modal` component and I wanted to always render its
+content into a specially-styled element at the root of the body (with `position:
+fixed`, say), I might write it like this:
+
+```ts
+import Component from '@glimmer/component';
+
+export default class Modal extends Component {
+  modalElement = document.getElemenyById('modal');
+}
+```
+
+```hbs
+{{#in-element modalElement}}
+  <h1>Modal</h1>
+  {{yield}}
+{{/in-element}}
+```
+
+Now in my app, I can invoke my `Modal` component as deep into the hierarchy as I
+want, and the content will be rendered into the root modal element:
+
+```hbs
+{{#if hasErrors}}
+  <Modal>
+    <p>You dun goofed:</p>
+    <ul>
+      {{#each errors as |error|}}
+        <li>{{error}}</li>
+      {{/each}}
+    </ul>
+  </Modal>
+{{/if}}
+```
+
+### Binary Templates
+
+It's crucial that web apps render instantly, or else users go elsewhere. When it
+comes to improving web performance, one of the most frequent recommendations
+you'll hear is to minimize the amount of total JavaScript in your app.
+
+There are two reasons for this: not only does more JavaScript take longer to
+download, just _parsing_ the JavaScript can become a noticeable bottleneck on
+underpowered devices.
+
+Complicating the advice to "use less JavaScript" is the fact that most modern
+JavaScript libraries, including Angular, React, Vue and Svelte, compile
+component templates (or JSX) to JavaScript, which gets embedded in application
+code. Without aggressive hand optimization, more templates means more
+JavaScript.
+
+Ember used to do the same thing, compiling Handlebars templates into JavaScript
+code that would first create and then update a component's DOM tree.
+
+With Glimmer, however, we took a different approach. Instead of generating
+JavaScript, today we compile templates into a JSON data structure of "opcodes,"
+or rendering instructions. A small runtime evaluates these opcodes, translating
+them into DOM creation, DOM updates, invocation of component hooks, etc.
+
+Not only is [a JSON parser much faster than a full-blown JavaScript
+parser](https://jsperf.com/json-parse-vs-eval-corrected/1), aggressively sharing
+code in the Glimmer VM generates less on-device memory pressure and allows
+JavaScript engines like V8 to more quickly generate JIT-optimized code.
+
+Best of all, our compact JSON format is significantly smaller than the
+equivalent compiled JavaScript. We received many reports of apps dropping 30-50%
+in total (post-gzip!) size after upgrading to Ember 2.10, the first version to
+use this JSON-based approach.
+
+As exciting as this was, we knew that JSON was not the final word in compactly
+and efficiently representing compiled templates.
+
+At runtime, Glimmer VM would gather the JSON for each template and compile them
```

Comment:
"would" introduces past tense where you have been writing in present tense. For example above "today we compile...".

Perhaps "At runtime, the Glimmer VM today in Ember gathers the JSON for each template and compiles them into a final, internal representation that is a large array of 32-bit integers." I think the rest is good.